src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs

```diff
@@ -3813,23 +3813,40 @@ bool shouldConvertToIDisposableBeforeTry(IOperation resource)
 
             void processResource(IOperation resource, ArrayBuilder<(IVariableDeclarationOperation, IVariableDeclaratorOperation)>? resourceQueueOpt)
             {
-                // When ResourceType is a non-nullable value type, the expansion is:
+                // When ResourceType is a non-nullable value type that implements IDisposable, the expansion is:
                 //
                 // {
                 //   ResourceType resource = expr;
                 //   try { statement; }
                 //   finally { ((IDisposable)resource).Dispose(); }
                 // }
                 //
+                // Otherwise, when ResourceType is a non-nullable value type that implements
+                // disposal via pattern, the expansion is:
+                //
+                // {
+                //   try { statement; }
+                //   finally { resource.Dispose(); }
+                // }
                 // Otherwise, when Resource type is a nullable value type or
-                // a reference type other than dynamic, the expansion is:
+                // a reference type other than dynamic that implements IDisposable, the expansion is:
                 //
                 // {
                 //   ResourceType resource = expr;
                 //   try { statement; }
                 //   finally { if (resource != null) ((IDisposable)resource).Dispose(); }
                 // }
                 //
+                // Otherwise, when Resource type is a nullable value type or
```

Comment:
>when Resource type is a nullable value type [](start = 30, length = 43)

Are we testing this scenario?

Reply:
Yes, in `UsingFlow_05` but actually, it doesn't appear to be correctly reporting what we do. It matches the comments in the Builder, but that _isnt_ how we lower the code, as you pointed out we unwrap the nullable, which we're not representing here. https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQBMQGoA+ABATAAgAUBYAKAG8z9r9sAWfAWQApGBPAEQEsoAHAeygRgAGzgB+fFwB2vAK4wk+YP34jlASjIB6AFQAeYDJQA+Xdsqka12gAZ8zAOJwY3PoOFjmG/OMkz5GC0rG2pLUNDgfABefBgEOTgAbioIgF9U6jS9fW0jaVNzMkymTh4BIVEJfGdXco8q7xiTOIALBH4Ad3xpORERFNIM0jJYBIBjGFK3Cs84fBBaAEYAOgBJGYaxMnCabABmWgZNqDgm6JaYdq6evoGyNKA=

Reply:
Oh, interesting. Actually we're representing in ControlFlowGraph what the spec says we *should* be doing: 
http://sourceroslyn.io/#Microsoft.CodeAnalysis.CSharp/Lowering/LocalRewriter/LocalRewriter_UsingStatement.cs,277

Except in lowering we purposefully deviate from the spec to improve perf. 

@AlekseyTs what do you think? Should the CFG represent the spirit of the code in the spec, or what we *actually* lower to? I guess I lean towards representing the spec? because like the async stuff we give a general representation of what it's lowered to, not the implementation specific details.

Reply:
It looks to me that UsingFlow_05 and the sharplab link is about non-pattern disposal and this change is not supposed to make any changes for these scenarios. I am fine with the current behavior of the builder for them. The question I have is about pattern-based disposal for nullable value types. This is what this comment is about. Are we testing the scenario? 

---
In reply to: [534599322](https://github.com/dotnet/roslyn/pull/49481#discussion_r534599322) [](ancestors = 534599322)

Reply:
It is not currently possible to perform pattern disposal on a nullable value types: see https://github.com/dotnet/roslyn/issues/34701 

I will add a test to show the error recovery scenario, and update the comment in CFG to correctly reflect that.